src/NexusMods.DataModel/ChunkedStreams/LightweightLRUCache.cs

```diff
@@ -0,0 +1,121 @@
+﻿namespace NexusMods.DataModel.ChunkedStreams;
+
+/// <summary>
+/// An extremely lightweight LRU cache that is not thread safe. All values are expected
+/// to be unused by the time the next method on this struct is called.
+/// </summary>
+public struct LightweightLRUCache<TK, TV>
+    where TK : IEquatable<TK>
+{
+    private readonly int _size;
+    private readonly TK[] _keys;
+    private readonly TV[] _values;
+    private int _count;
+
+    /// <summary>
+    /// Creates a new cache with the given size.
+    /// </summary>
+    /// <param name="size"></param>
+    public LightweightLRUCache(int size)
+    {
+        _size = size;
+        _keys = new TK[size];
+        _values = new TV[size];
+        _count = 0;
+    }
+
+    /// <summary>
+    /// Gets the number of items in the cache.
+    /// </summary>
+    public int Count => _count;
+
+    /// <summary>
+    /// Tries to get a value from the cache. If the value is not in the cache, the value is set to default.
+    /// </summary>
+    /// <param name="key"></param>
+    /// <param name="value"></param>
+    /// <returns></returns>
+    public bool TryGet(TK key, out TV? value)
+    {
+        if (Find(key) is var index && index != -1)
+        {
+            value = _values[index];
+            Touch(index);
+            return true;
+        }
+        value = default;
+        return false;
+    }
+
+    /// <summary>
+    /// Moves the given index to the front of the cache.
+    /// </summary>
+    /// <param name="index"></param>
+    private void Touch(int index)
+    {
+        if (index == 0) return;
+        var key = _keys[index];
+        var val = _values[index];
+        Array.Copy(_keys, 0, _keys, 1, index);
```

Comment:
Avoid `Array.Copy` in this code. Because that code is non-generic, it needs to do some extra pre-processing; and we're dealing with small cache sizes.

Use `Buffer.MemMove` private API.

You can either do it with `Buffer.MemoryCopy();` or `_keys.AsSpanFast().CopyTo()`.

Reply:
Where are these private API methods at? I can seem to find them.

Reply:
`Buffer.MemMove` is internal to the runtime. It's basically `memcpy`.

`Buffer.MemoryCopy();` and `_keys.AsSpanFast().CopyTo().` would be two efficient ways to access that method with minimal overhead.

Not sure which one's faster, the second one technically avoids pinning.

Note: `AsSpanFast` needs is an extension used in some of our other projects (e.g. Paths), just grab it from there.